### DISCUSSION TITLE
discuss: fetch data from fmarket

### DISCUSSION THREAD

**Discussion by h7b** [vietfin/vietfin#1] [Ideas]
Mình bắt đầu bằng 1 trường hợp mà a H đang viết.

a H muốn cào dữ liệu về thông tin các mutual funds được giao dịch thông qua [fmarket](https://fmarket.vn/home)

Dùng dev tool trong Chrome sẽ dò được public API của fmarket.

Để xem thử nội dung của gói tin trả về thì Ti vào [reqbin](https://reqbin.com/)

- mục `api` thì điền vào `https://api.fmarket.vn/res/products/filter`
- chọn method là `POST`
- ở phần `JSON` payload thì dán đoạn này vào

```json
{
        "types": ["NEW_FUND", "TRADING_FUND"],
        "issuerIds": [],
        "sortOrder": "DESC",
        "sortField": "navTo6Months",
        "page": 1,
        "pageSize": 100,
        "isIpo": "False",
        "fundAssetTypes": [],
        "bondRemainPeriods": [],
        "searchField": "",
        "isBuyByReward": "False",
        "thirdAppIds": []
    }
```

Xong rồi bấm `Send`

Kết quả trả về sẽ nằm ở bảng `Content` nếu `Status` là `200`

Cái chúng ta cần làm với VietFin package là viết hàm chuyển đổi cấu trúc và nội dung gói tin trả về sang 1 tabular format (dạng dataframe chẳng hạn) và output cái đó ra cho người dùng cuối. Tương tự như cách `yfinance` cào dữ liệu của Yahoo Finance.

a H đang thử nghiệm code ý tưởng mẫu trên 1 file `ipynb`, chứ chưa tích hợp vào thành hàm trong package.

**Comment by myadang04** [comment]
ở bên tụi em dùng tool Postman (https://www.postman.com/) để fetch data from API á, nhìn nó professional với đỡ rối hơn nên a Huy thử xài nha

**Comment by myadang04** [comment]
do em không am hiểu lắm về data mình đang có là gì, a Huy nói cụ thể hơn output dạng tabular của mình sẽ cần những info (header) gì, và mình sẽ làm gì với output đó không ạ?

**Reply by h7b** [reply]
@myadang04 hi Ti (aka "T")

a H push lên `dev` branch cái [demo notebook](https://github.com/h7b/vietfin/blob/dev/examples/idea_notebook.ipynb). T xem thử có trả lời được thắc mắc ko.

input data m đang có là từ việc cào dữ liệu của các API từ các công ty chứng khoán / nền tảng giao dịch ở Vietnam
- a H chọn khởi đầu với cái API của nhà cung cấp này vì họ tương đối dễ chịu nhất và chưa giới hạn lượt / thời gian truy cập gì nhiều
- T muốn xem cấu trúc json file from API request thì đọc notebook. Ko thì dùng postman (trong notebook a H có để api url, payload, headers) hoặc reqbin như a H cũng được

Nói về output:
- vd như có 1 nhóm thông tin là giá bán của chứng chỉ quỹ (timeseries data) và  a H muốn trích xuất thông tin này từ json response from API call
- output mong đợi sẽ là 1 pandas dataframe
- column label của dataframe này thì m sẽ có thể đặt là ['date', 'price', 'update_at'] chẳng hạn

Tại sao lại chọn output là 1 pandas df?
- vì hiện giờ thói quen của nhiều analyst vẫn là dùng package `pandas` cho việc xử lý dữ liệu dạng bảng
- nên xuất ra pandas df thì đỡ mất công cho người dùng ít kinh nghiệm chuyển đổi sang dạng khác
- các output format khả dĩ mà a H đã tìm hiểu: `polars` dataframe, serializable data (e.g. json) 

Và m có thể làm gì với output này?
- cái này thì tùy thuộc vào nhu cầu của người phân tích thôi
- họ có thể dựa vào dữ liệu giá của chứng chỉ quỹ để xem có correlation with market movement (e.g. vnindex), hoặc so sánh hiệu suất hoạt động giữa các quỹ để chọn quỹ mà họ muốn bỏ tiền vào
- cơ bản thì package VietFin chỉ là 1 python interface for publicly available API, so users can fetch data and work on it. The use cases are up to the sky limit

Trong demo notebook thì a H liệt kê ra danh sách tất cả các quỹ đang bán thông qua nền tảng giao dịch Fmarket.

Còn khá nhiều nhóm thông tin mà m cần khai thác. Và các API từ nhà cung cấp khác nhau sẽ cần cách cào / xử lý dữ liệu khác nhau.

Khi chúng ta định hình được cách hoạt động và nguyên lý thiết kế rồi những việc sau đó có lẽ chỉ là lao động chân tay ko cần dùng não :))

**Reply by myadang04** [reply]
Oh dạ em hiểu rùi, sorry a Huy em mới đi học lại nên schedule hơi rối, trả lời sẽ hơi chậm nha huhu. 
Em không rành về stock market lắm nên A Huy có thêm tài liệu gì về cái này (họ có thể dựa vào dữ liệu giá của chứng chỉ quỹ để xem có correlation with market movement (e.g. vnindex), hoặc so sánh hiệu suất hoạt động giữa các quỹ để chọn quỹ mà họ muốn bỏ tiền vào) để em hiểu thêm output của mình hong ạ?
Và giờ là mình sẽ bắt đầu code để ra những pandas table luôn đúng hong ạ?

**Reply by h7b** [reply]
@myadang04 

Ý T chưa hiểu output là data format hay là gì của ouput nhỉ?
- nếu là format thì cơ bản a H định sẽ trả về 1 pandas dataframe, hoặc 1 object dạng dictionary
- nếu là data structure của response from api call thì T xem trong [demo notebook](https://github.com/h7b/vietfin/blob/dev/examples/idea_notebook.ipynb), `data = response.json()` là phần nhận dữ liệu trả về từ API
- nếu là ko biết chọn thông tin (cột dữ liệu) gì từ api để xuất ra dữ liệu thì cái này xem a H hay T cần cái gì thì chọn xuất ra cái đó. Sau này khi a H publish first version lên pypi, nếu có người dùng khác nữa thích thay đổi thì có thể pull request để mình thêm vào

Còn example usage thì có mấy cái này a H đã xem qua
- [tutorial](https://www.youtube.com/watch?v=1O_BenficgE) này của dataquest minh họa cách dùng dữ liệu giá ohlc lịch sử để dự đoán giá cổ phiếu bằng machine learning cơ bản. Cái bước lấy giá dữ liệu ohlc là thông qua package `yfinance`. Cơ bản thì `vietfin` cũng sẽ đóng vai trò tương tự như `yfinance`, nhưng là dành riêng cho dữ liệu của thị trường Vietnam.
- Vietnam là thị trường cận biên (ko được xếp hạng, đánh giá nghiêm minh) như các thị trường phát triển như Nasdaq của Mỹ hay Kospi của Hàn, nên dữ liệu trên `yfinance` đôi khi sẽ thiếu / ko có. Vì vậy mình phát triển `vietfin` package để cho vui :D
- [paper](https://research-information.bris.ac.uk/en/publications/fundamental-analysis-and-the-use-of-financial-statement-informati) này của 1 nhóm tác giả người Việt đăng lên ở Anh, có vẻ là sinh viên vì a H thấy nội dung viết a H đọc hiểu được. Paper viết về tính hiệu quả của fundamental analysis in frontier market (Vietnam). Trong paper thì họ dùng nguồn dữ liệu từ Refinitiv (nghe nói annual subscription 30k USD). Nên sau khi bổ sung để `vietfin` cào được phần dữ liệu về nền tảng doanh nghiệp thì a H định sẽ reproduce cái bài báo này để làm ví dụ minh họa cho ứng dụng của `vietfin`

Nếu T ko khoái mấy cái lao động tay chân nhàm chán như api parsing kiểu này thì nhảy qua xem giúp a H về cách thiết kế cấu trúc phần mềm ở post #2   này thì càng tốt á.
